From: Daniel van Vugt <daniel.van.vugt@canonical.com>
Date: Tue, 11 Jul 2023 18:42:38 +0800
Subject: backends: Handle null views instead of creating dummy views

If view initialization fails then don't add the view, rather than
adding a dummy offscreen view. This avoids flooding the log with
offscreen frame clock confusion:

Before:
```
libmutter-WARNING **: 15:47:27.763: Failed to allocate onscreen framebuffer for /dev/dri/card0: Failed to allocate surface: Function not implemented
Clutter-WARNING **: 15:47:28.557: (../clutter/clutter/clutter-frame-clock.c:419):clutter_frame_clock_notify_presented: code should not be reached
Clutter-WARNING **: 15:47:28.563: (../clutter/clutter/clutter-frame-clock.c:419):clutter_frame_clock_notify_presented: code should not be reached
Clutter-WARNING **: 15:47:28.567: (../clutter/clutter/clutter-frame-clock.c:419):clutter_frame_clock_notify_presented: code should not be reached
(repeats forever)

```

After:
```
libmutter-WARNING **: 16:09:04.945: Failed to create view for Unknown 46" on None-1: Failed to allocate onscreen framebuffer for /dev/dri/card0: Failed to allocate surface: Function not implemented
```

Bug-Ubuntu: https://launchpad.net/bugs/1967707
Origin: Upstream commit: e951265806510ed3a9c493a04977b8760a7f69e3
Forwarded: yes
---
 src/backends/meta-renderer.c                       | 32 ++++++++---
 src/backends/meta-renderer.h                       |  9 +--
 src/backends/native/meta-renderer-native.c         | 67 +++++++---------------
 src/backends/x11/nested/meta-renderer-x11-nested.c |  9 +--
 4 files changed, 55 insertions(+), 62 deletions(-)

diff --git a/src/backends/meta-renderer.c b/src/backends/meta-renderer.c
index 9faf651..1a7a40b 100644
--- a/src/backends/meta-renderer.c
+++ b/src/backends/meta-renderer.c
@@ -92,15 +92,24 @@ meta_renderer_create_cogl_renderer (MetaRenderer *renderer)
 }
 
 static MetaRendererView *
-meta_renderer_create_view (MetaRenderer       *renderer,
-                           MetaLogicalMonitor *logical_monitor,
-                           MetaOutput         *output,
-                           MetaCrtc           *crtc)
+meta_renderer_create_view (MetaRenderer        *renderer,
+                           MetaLogicalMonitor  *logical_monitor,
+                           MetaOutput          *output,
+                           MetaCrtc            *crtc,
+                           GError             **error)
 {
-  return META_RENDERER_GET_CLASS (renderer)->create_view (renderer,
+  MetaRendererView *view;
+
+  view = META_RENDERER_GET_CLASS (renderer)->create_view (renderer,
                                                           logical_monitor,
                                                           output,
-                                                          crtc);
+                                                          crtc,
+                                                          error);
+
+  if (view)
+    meta_renderer_add_view (renderer, view);
+
+  return view;
 }
 
 /**
@@ -127,9 +136,16 @@ create_crtc_view (MetaLogicalMonitor *logical_monitor,
 {
   MetaRenderer *renderer = user_data;
   MetaRendererView *view;
+  g_autoptr (GError) error = NULL;
 
-  view = meta_renderer_create_view (renderer, logical_monitor, output, crtc);
-  meta_renderer_add_view (renderer, view);
+  view = meta_renderer_create_view (renderer, logical_monitor, output, crtc, &error);
+  if (!view)
+    {
+      g_warning ("Failed to create view for %s on %s: %s",
+                 meta_monitor_get_display_name (monitor),
+                 meta_output_get_name (output),
+                 error->message);
+    }
 }
 
 static void
diff --git a/src/backends/meta-renderer.h b/src/backends/meta-renderer.h
index 66cb71d..2e3e7cd 100644
--- a/src/backends/meta-renderer.h
+++ b/src/backends/meta-renderer.h
@@ -38,10 +38,11 @@ struct _MetaRendererClass
   GObjectClass parent_class;
 
   CoglRenderer * (* create_cogl_renderer) (MetaRenderer *renderer);
-  MetaRendererView * (* create_view) (MetaRenderer       *renderer,
-                                      MetaLogicalMonitor *logical_monitor,
-                                      MetaOutput         *output,
-                                      MetaCrtc           *crtc);
+  MetaRendererView * (* create_view) (MetaRenderer        *renderer,
+                                      MetaLogicalMonitor  *logical_monitor,
+                                      MetaOutput          *output,
+                                      MetaCrtc            *crtc,
+                                      GError             **error);
   void (* rebuild_views) (MetaRenderer *renderer);
   void (* resume) (MetaRenderer *renderer);
   GList * (* get_views_for_monitor) (MetaRenderer *renderer,
diff --git a/src/backends/native/meta-renderer-native.c b/src/backends/native/meta-renderer-native.c
index 3c22b4e..24c12f8 100644
--- a/src/backends/native/meta-renderer-native.c
+++ b/src/backends/native/meta-renderer-native.c
@@ -1328,32 +1328,12 @@ should_force_shadow_fb (MetaRendererNative *renderer_native,
   return meta_kms_device_prefers_shadow_buffer (kms_device);
 }
 
-static CoglFramebuffer *
-create_fallback_offscreen (MetaRendererNative *renderer_native,
-                           int                 width,
-                           int                 height)
-{
-  CoglOffscreen *fallback_offscreen;
-  GError *error = NULL;
-
-  fallback_offscreen = meta_renderer_native_create_offscreen (renderer_native,
-                                                              width,
-                                                              height,
-                                                              &error);
-  if (!fallback_offscreen)
-    {
-      g_error ("Failed to create fallback offscreen framebuffer: %s",
-               error->message);
-    }
-
-  return COGL_FRAMEBUFFER (fallback_offscreen);
-}
-
 static MetaRendererView *
-meta_renderer_native_create_view (MetaRenderer       *renderer,
-                                  MetaLogicalMonitor *logical_monitor,
-                                  MetaOutput         *output,
-                                  MetaCrtc           *crtc)
+meta_renderer_native_create_view (MetaRenderer        *renderer,
+                                  MetaLogicalMonitor  *logical_monitor,
+                                  MetaOutput          *output,
+                                  MetaCrtc            *crtc,
+                                  GError             **error)
 {
   MetaRendererNative *renderer_native = META_RENDERER_NATIVE (renderer);
   MetaBackend *backend = meta_renderer_get_backend (renderer);
@@ -1374,7 +1354,7 @@ meta_renderer_native_create_view (MetaRenderer       *renderer,
   MtkRectangle view_layout;
   MetaRendererViewNative *view_native;
   EGLSurface egl_surface;
-  GError *error = NULL;
+  GError *local_error = NULL;
 
   crtc_config = meta_crtc_get_config (crtc);
   crtc_mode_info = meta_crtc_mode_get_info (crtc_config->mode);
@@ -1388,15 +1368,13 @@ meta_renderer_native_create_view (MetaRenderer       *renderer,
 
       if (!meta_renderer_native_ensure_gpu_data (renderer_native,
                                                  gpu_kms,
-                                                 &error))
+                                                 &local_error))
         {
-          g_warning ("Failed to create secondary GPU data for %s: %s",
-                      meta_gpu_kms_get_file_path (gpu_kms),
-                      error->message);
-          use_shadowfb = FALSE;
-          framebuffer = create_fallback_offscreen (renderer_native,
-                                                   onscreen_width,
-                                                   onscreen_height);
+          g_propagate_prefixed_error (error, local_error,
+                                      "Failed to create secondary GPU data for %s: ",
+                                      meta_gpu_kms_get_file_path (gpu_kms));
+
+          return NULL;
         }
       else
         {
@@ -1410,15 +1388,12 @@ meta_renderer_native_create_view (MetaRenderer       *renderer,
                                                       onscreen_width,
                                                       onscreen_height);
 
-          if (!cogl_framebuffer_allocate (COGL_FRAMEBUFFER (onscreen_native), &error))
+          if (!cogl_framebuffer_allocate (COGL_FRAMEBUFFER (onscreen_native), &local_error))
             {
-              g_warning ("Failed to allocate onscreen framebuffer for %s: %s",
-                         meta_gpu_kms_get_file_path (gpu_kms),
-                         error->message);
-              use_shadowfb = FALSE;
-              framebuffer = create_fallback_offscreen (renderer_native,
-                                                       onscreen_width,
-                                                       onscreen_height);
+              g_propagate_prefixed_error (error, local_error,
+                                          "Failed to allocate onscreen framebuffer for %s: ",
+                                          meta_gpu_kms_get_file_path (gpu_kms));
+              return NULL;
             }
           else
             {
@@ -1438,9 +1413,9 @@ meta_renderer_native_create_view (MetaRenderer       *renderer,
       virtual_onscreen = meta_renderer_native_create_offscreen (renderer_native,
                                                                 onscreen_width,
                                                                 onscreen_height,
-                                                                &error);
+                                                                &local_error);
       if (!virtual_onscreen)
-        g_error ("Failed to allocate back buffer texture: %s", error->message);
+        g_error ("Failed to allocate back buffer texture: %s", local_error->message);
       use_shadowfb = FALSE;
       framebuffer = COGL_FRAMEBUFFER (virtual_onscreen);
     }
@@ -1468,9 +1443,9 @@ meta_renderer_native_create_view (MetaRenderer       *renderer,
       offscreen = meta_renderer_native_create_offscreen (renderer_native,
                                                          offscreen_width,
                                                          offscreen_height,
-                                                         &error);
+                                                         &local_error);
       if (!offscreen)
-        g_error ("Failed to allocate back buffer texture: %s", error->message);
+        g_error ("Failed to allocate back buffer texture: %s", local_error->message);
     }
 
   if (meta_backend_is_stage_views_scaled (backend))
diff --git a/src/backends/x11/nested/meta-renderer-x11-nested.c b/src/backends/x11/nested/meta-renderer-x11-nested.c
index 2e05463..974143e 100644
--- a/src/backends/x11/nested/meta-renderer-x11-nested.c
+++ b/src/backends/x11/nested/meta-renderer-x11-nested.c
@@ -61,10 +61,11 @@ create_offscreen (CoglContext *cogl_context,
 }
 
 static MetaRendererView *
-meta_renderer_x11_nested_create_view (MetaRenderer       *renderer,
-                                      MetaLogicalMonitor *logical_monitor,
-                                      MetaOutput         *output,
-                                      MetaCrtc           *crtc)
+meta_renderer_x11_nested_create_view (MetaRenderer        *renderer,
+                                      MetaLogicalMonitor  *logical_monitor,
+                                      MetaOutput          *output,
+                                      MetaCrtc            *crtc,
+                                      GError             **error)
 {
   MetaBackend *backend = meta_renderer_get_backend (renderer);
   ClutterBackend *clutter_backend = meta_backend_get_clutter_backend (backend);
